### PR TITLE
[CGKIELE-107] mallet

### PR DESCRIPTION
**Pull request by rtkaczyk**
The minimalist CLI wallet for KEVM/IELE testnet.

See [the list of missing features](https://iohk.myjetbrains.com/youtrack/issue/CGKIELE-107#comment=93-19538)

Additional improvements:
- a fix for `sbt-verify`: to regenerate checksums simply type `verifyGenerate`. This will create a file `verify.sbt` in the root dir - no need for manual copying of the output

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -143,7 +148,14 @@ val verifyDeps = Seq(
...
   "org.objenesis" % "objenesis" sha1 "272bab9a4e5994757044d1fc43ce480c8cb907a4",
-  "org.hamcrest" % "hamcrest-all" sha1 "63a21ebc981131004ad02e0434e799fd7f3a8d5a"
+  "org.hamcrest" % "hamcrest-all" sha1 "63a21ebc981131004ad02e0434e799fd7f3a8d5a",
+
+  // mallet
+  "com.github.scopt" % "scopt" sha1 "e078455e1a65597146f8608dab3247bf1eb92e6e",
+  "org.jline" % "jline" sha1 "dfb4e9e15e981634155ce063fa697b2b8964d507",
+  "org.scala-lang.modules" % "scala-parser-combinators" sha1 "3c1c5475ece77c41e18dd971f8f818c091e4961c",
+  // FIXME: WTF?
+  "com.google.j2objc" % "j2objc-annotations" sha1 "976d8d30bebc251db406f2bdb3eb01962b5685b3"
```
Comment (loverdos) [inline build.sbt]: I did a bit of searching. This is a dependency pulled in via guava `v22` (check image below) and is discussed [here](https://github.com/google/guava/issues/2824).


selection_108

Comment (rtkaczyk) [reply build.sbt]: What uses guava?

Comment (loverdos) [reply build.sbt]: Using `sbt` with `addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.9.0")`:
```
$ sbt whatDependsOn com.google.guava guava 22.0
[info] com.google.guava:guava:22.0
[info]   +-io.atomix:atomix-utils:2.1.0-beta1
[info]   | +-io.atomix:atomix-cluster:2.1.0-beta1
[info]   | [...]   
[info]   +-org.consensusresearch:scrypto_2.12:1.2.0-RC3 [S]
[info]   | [...]
[info]   +-org.scorexfoundation:iodb_2.12:0.3.0 [S]
[info]     
```

File: src/main/scala/io/iohk/ethereum/consensus/ethash/EthashMiner.scala [hunk 2]
```diff
@@ -93,7 +93,7 @@ class EthashMiner(
         val startTime = System.currentTimeMillis()
         val mineResult = mine(headerHash, block.header.difficulty.toLong, dagSize, dag, miningConfig.mineRounds)
         val time = System.currentTimeMillis() - startTime
-        val hashRate = (mineResult.triedHashes * 1000) / time
+        val hashRate = if (time > 0) (mineResult.triedHashes * 1000) / time else Long.MaxValue
```
Comment (loverdos) [inline src/main/scala/io/iohk/ethereum/consensus/ethash/EthashMiner.scala]: I was wondering if we could use a "better" (=monotonic) clock, e.g. `System.nanoTime()` whose purpose is to measure elapsed time. There is no strong guarantee that this is indeed going to be monotonic because it ultimately depends on the OS facilities but this is generally not a problem in Linux nowadays. 

Another possible route is to not update the hash rate if we cannot reliably compute it.

Comment (loverdos) [reply src/main/scala/io/iohk/ethereum/consensus/ethash/EthashMiner.scala]: Note: guava has a `com.google.common.base.StopWatch` class for this purpose.

File: src/main/scala/io/iohk/ethereum/mallet/common/Util.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  def exceptionToString(ex: Throwable): String = {
+    val sw = new StringWriter()
+    sw.append(ex.getMessage + "\n")
+    ex.printStackTrace(new PrintWriter(sw))
+    sw.toString
+  }
+
+  implicit class OptionOps[T](opt: Option[T]) {
+    def toEither[U](left: U): Either[U, T] =
```
Comment (loverdos) [inline src/main/scala/io/iohk/ethereum/mallet/common/Util.scala]: There is `scala.Option.toRight`, so we can avoid introducing a new construct.

File: src/main/scala/io/iohk/ethereum/consensus/ethash/EthashMiner.scala (hunk 2 above)
Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/consensus/ethash/EthashMiner.scala]: I'd rather not spend too much time on this, this is not really a big problem anyway. I can add a FIXME though

File: src/main/scala/io/iohk/ethereum/mallet/common/Util.scala (hunk 3 above)
Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/mallet/common/Util.scala]: lol, how did I miss that? :)

File: build.sbt (hunk 1 above)
Comment (rtkaczyk) [reply build.sbt]: @loverdos I removed the `j2objc-annotations` artifact from my local cache. After rebuilding, the checksum changed and it passed the verification. See also the updated PR description

File: build.sbt [hunk 4]
```diff
@@ -161,15 +73,16 @@ val root = project.in(file("."))
...
+      verifyOutputFile in verifyGenerate := baseDirectory.value / "verify.sbt",
       verifyOptions in verify := VerifyOptions(
         includeBin = true,
         includeScala = true,
         includeDependency = true,
         excludedJars = Nil,
-        warnOnUnverifiedFiles = true,
-        warnOnUnusedVerifications = true
+        warnOnUnverifiedFiles = false,
+        warnOnUnusedVerifications = false
```
Comment (LukasGasior1) [inline build.sbt]: why this change? it seems keeping it true is more secure

File: src/main/scala/io/iohk/ethereum/mallet/interpreter/AST.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  /**
+    * Base trait for literal types of arguments from user input.
+    * Note: these are different and require mapping to command parameter types
+    */
+  sealed trait Literal {
+    def input: String
+  }
+
+  case class Quoted(input: String) extends Literal {
+    def unqoute: String = unquote(input)
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/mallet/interpreter/AST.scala]: typo: unqoute -> unquote

File: src/universal/bin/mantis [hunk 6]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+./bin/mantis-app -Dconfig.file=./conf/mantis.conf -Dlogback.configurationFile=./conf/logback.xml -J-Xss10M
```
[the remark this answers is not in the document]
Comment (rtkaczyk) [reply src/universal/bin/mantis]: Good point, I was actually going to explore another way with those scripts, thanks for reminding me.

File: src/main/scala/io/iohk/ethereum/mallet/interpreter/Parameter.scala [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      case q: Quoted => ByteString(q.unqoute)
+    }
+  }
+
+  case object Addr20 extends ParamType {
+    type T = Address
+
+    protected val convert = {
+      case h: Hex if h.bytes.nonEmpty && h.bytes.length <= 20 =>
+        Address(h.bytes)
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/mallet/interpreter/Parameter.scala]: can we also support quoted addresses?

File: build.sbt (hunk 4 above)
Comment (rtkaczyk) [reply build.sbt]: `false` means error instead of warning. I actually didn't plan on changing this, but it happened during some fixes I applied wrt `sbt-verify`. 

I'm not having any problems with `sbt-verify` now, so I wouldn't mind staying with `false`. WDYT?

File: src/main/scala/io/iohk/ethereum/mallet/interpreter/Parameter.scala (hunk 7 above)
Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/mallet/interpreter/Parameter.scala]: Why?

File: build.sbt (hunk 4 above)
Comment (LukasGasior1) [reply build.sbt]: Makes sense, I didn't know it means an error :)

File: src/main/scala/io/iohk/ethereum/mallet/interpreter/Parameter.scala (hunk 7 above)
Comment (LukasGasior1) [reply src/main/scala/io/iohk/ethereum/mallet/interpreter/Parameter.scala]: My first intuition was to try with quotes. Also geth console uses quoted addresses (it doesn't even accept unquoted ones)

File: verify.sbt [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  "io.netty" % "netty-transport-native-unix-common" sha1 "fee46a4e2b4f2f096532ca443f4f63e08b205318",
+  "io.atomix" % "atomix-raft" sha1 "e45af08b70220dd010d6f193e3d38147b4be35fe",
+  "io.atomix" % "atomix-primitive" sha1 "73d9bc7c859856832178afd271f1a573e6a7c7e6",
+  "io.atomix" % "atomix-storage" sha1 "136f0b221acbc2680f099b8ff3a34f8cc1592fe7",
+  "io.atomix" % "atomix-primary-backup" sha1 "1c895965e3e67a152ffbccb4283b6cee91b4ea61",
+  "io.netty" % "netty-tcnative-boringssl-static" sha1 "ff5f2d6db5aaa1b4df1b381382cd6581844aad9d",
+  "com.github.scopt" % "scopt" sha1 "e078455e1a65597146f8608dab3247bf1eb92e6e"
+)
+
+verifyOptions in verify := VerifyOptions(
```
Comment (LukasGasior1) [inline verify.sbt]: is this needed here? how does it relate to the options in build.sbt?

Comment (rtkaczyk) [reply verify.sbt]: This file is auto-generated, see the PR description

**Comment by LukasGasior1** [comment]
One thing I noticed: we probably don't want to store the `importPrivateKey` command in the history

**Comment by rtkaczyk** [comment]
> importPrivateKey command in the history

That's a really good point!

File: src/main/scala/io/iohk/ethereum/mallet/interpreter/Parameter.scala (hunk 7 above)
Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/mallet/interpreter/Parameter.scala]: I guess it would be easy to do, but for now I wouldn't go further than adding a `TODO` comment to consider this.
Firstly, I'd rather have the syntax be strict, and secondly there's supporting account aliases (which would be quoted) on the list missing features.

**Comment by LukasGasior1** [comment]
Another one: `-d` flag does not work:
```
ubuntu@ip-10-0-1-82:~/mantis-faucet/mantis-1.0-daedalus-rc1$ ./bin/mallet -d "/home/ubuntu/keystore" "http://10.0.1.82:8546"                                                                                                               
[residual] arg = '/home/ubuntu/keystore'                                                                                                                                                                                                  
[residual] arg = 'http://10.0.1.82:8546'
# Executing command line:
java
-Dconfig.file=./conf/mantis.conf                                                                                                                                                                                                          
-Dlogback.configurationFile=./conf/logback.xml                                                                                                                                                                                            
-Xss10M
-Dconfig.file=./conf/mallet.conf                                                                                                                                                                                                          
-cp                                                                                                                                                                                                                                       

-jar                                                                                                                                                                                                                                      
/home/ubuntu/mantis-faucet/mantis-1.0-daedalus-rc1/lib/mantis.mantis-1.0-daedalus-rc1-launcher.jar                                                                                                                                        
mallet                                                                                                                                                                                                                                    
/home/ubuntu/keystore
http://10.0.1.82:8546

Error: node URI scheme must be explicit and either 'http' or 'https'                                                                                                                                                                      
Error: Unknown argument 'http://10.0.1.82:8546'                                                                                                                                                                                           
Try --help for more information.                               
```

full `--data-dir` works just fine

**Comment by rtkaczyk** [comment]
@LukasGasior1 `-d` was conflicting with the launcher's script debug option, but adding `--` fixed the issue.

Also on the subject, note there are no longer separate `mantis` and `mantis-app` scripts. Turns out that adding `-Dconfig.file=...` in the mallet script overrides the defaults from `application.ini` (which was the original problem that the hack tried to solve)

In other news:
- `importPrivateKey` is no longer saved in history
- absence of a receipt is correctly handled
- receipts are displayed as formatted JSON

File: src/main/scala/io/iohk/ethereum/mallet/main/OptionParser.scala [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+  private implicit val addrRead: scopt.Read[Address] =
+    scopt.Read.reads { str =>
+      val bytes = StringUtil.hexToBytes(str)
+      require(bytes.nonEmpty && bytes.length <= Address.Length, "invalid address length")
+      Address(bytes)
+    }
+
+  def apply(args: Seq[String]): Option[ClOptions] = {
+    val parser = new scopt.OptionParser[ClOptions]("scopt") {
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/mallet/main/OptionParser.scala]: super minor: the string argument here is a `programName`, guess it should be `Constants.AppName` rather than `"scopt"`

Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/mallet/main/OptionParser.scala]: Not that minor, good catch

File: src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+import scala.util.{Failure, Success, Try}
+
+object RpcClient {
+  // TODO: CL option to enable akka logging
+  private val akkaConfig = ConfigFactory.load("mallet")
+
+  implicit val actorSystem = ActorSystem("mallet_rpc", akkaConfig)
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala]: why not instantiate it in Mallet app and pass to RpcClient as a parameter?

File: src/main/scala/io/iohk/ethereum/mallet/main/ConstPasswordReader.scala [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package io.iohk.ethereum.mallet.main
+
+import io.iohk.ethereum.mallet.service.PasswordReader
+
+/**
+  * This implementation of [[PasswordReader]] is in non-interactive mode when the password is
```
[the remark this answers is not in the document]
Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/mallet/main/ConstPasswordReader.scala]: Isn't it a bug? The `PasswordReader` is imported. See [this scaladoc](https://github.com/input-output-hk/mantis/blob/phase/iele_testnet/src/main/scala/io/iohk/ethereum/domain/Blockchain.scala#L28)

File: build.sbt [hunk 12]
```diff
@@ -186,7 +96,6 @@ scalacOptions := Seq(
   "-unchecked",
   "-deprecation",
   "-feature",
-  "-Xfatal-warnings",
   "-Xlint:unsound-match",
```
[the remark this answers is not in the document]
Comment (rtkaczyk) [reply build.sbt]: I can re-enable it. I always disable it when working on something complicated, because of unused imports. I'm adding and removing code back and forth, and going to imports for each such change kills my concentrations and infuriates me.

File: src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala (hunk 10 above)
Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala]: I meant to avoid `ActorSystem` leaking out of `RpcClient` cause it's an internal detail. Obviously this is not the case now, because I forgot to add `shutdown()` method to `RpcClient` - then the actor system could be private to `RpcClient`. What do you think about  such approach?

Comment (LukasGasior1) [reply src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala]: To me, actor system is not really internal to rpc client. Even though for mallet it's only used here, it has an impact on the whole app (for example, keeps it alive until it's shutdown), that's why I'd opt for putting in on app-level and pass (perhaps implicitly) to rpc client. It would also make RpcClient's dependencies more clear as well as make it reusable.

Comment (LukasGasior1) [reply src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala]: also, is there any reason why we're using the `actorSystem.dispatcher` instead of global EC or, even better, implicit ec passed as a constructor param?

Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala]: What's wrong with using `dispatcher`?

Comment (LukasGasior1) [reply src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala]: since you're using blocking operation (await), doesn't that mean that one of the threads that akka uses internally will be blocked? I know, in this case it doesn't matter since RpcClient is the only component using actors, but for the reasons mentioned above (reusability) I think it would be better to let the user of this class choose which EC he prefers (or, alternatively use global ec that won't block system dispatcher's thread).

File: src/main/scala/io/iohk/ethereum/mallet/main/ConstPasswordReader.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/mallet/main/ConstPasswordReader.scala]: @loverdos wdyt?

[the remark this answers is not in the document]
Comment (loverdos) [reply src/main/scala/io/iohk/ethereum/mallet/main/ConstPasswordReader.scala]: It works only if the referenced thing is in the same package, not if it is imported. I do not know if this is intended behaviour or a bug though. So I just use the FQN on these occasions.

File: src/main/resources/mallet.conf [hunk 13]
```diff
@@ -0,0 +1,28 @@
+akka {
+  # to enable logging use: ["akka.event.slf4j.Slf4jLogger"]
+  loggers = []
+
+  loglevel = OFF
+
+  ssl-config {
+    trustManager = {
+      stores = [
+        # TODO: move to Wiki maybe?
```
[the remark this answers is not in the document]
Comment (rtkaczyk) [reply src/main/resources/mallet.conf]: As discussed, it will be done in #445

File: build.sbt (hunk 12 above)
[the remark this answers is not in the document]
Comment (loverdos) [reply build.sbt]: There is a keystroke in IntelliJ that takes care of unused imports :)

File: src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+object RpcClient {
+
+  /**
+    * This factory method is defining an ActorSystem, ActorMaterializer and ExecutionContext for
+    * the [[RpcClient]]. To customize these dependencies use [[RpcClient]]'s constructor
+    */
+  def apply(node: Uri): RpcClient = {
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala]: @LukasGasior1 hopefully this should address your remarks about reusability, while limiting the scope of dependencies in mallet.

Comment (LukasGasior1) [reply src/main/scala/io/iohk/ethereum/mallet/service/RpcClient.scala]: ok, let's leave it that way
